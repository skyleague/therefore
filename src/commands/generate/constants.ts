import { name } from '../../../package.json'

export const generatedBy = `Generated by ${name}`
export const generatedAjv = `Generated by Ajv https://ajv.js.org/guide/managing-schemas.html#standalone-validation-code`
