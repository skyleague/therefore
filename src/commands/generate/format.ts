import { exec } from 'node:child_process'
import { promisify } from 'node:util'
import { memoize } from '@skyleague/axioms'
import type { ThereforeOutputType } from '../../lib/output/types.js'

// biome-ignore lint/suspicious/noExplicitAny: it's a third-party library
export type Prettier = any
export async function maybeLoadPrettier() {
    try {
        // @ts-ignore
        return await import('prettier')
    } catch {
        return undefined
    }
}

export async function formatContent({
    prettier,
    input,
    file,
    type,
}: {
    prettier: Prettier
    input: string
    file: string
    type: ThereforeOutputType
}): Promise<string> {
    if (prettier !== undefined) {
        const options = await prettier.resolveConfig(file)
        return prettier.format(input, {
            ...(options ?? {}),
            parser: type,
        })
    }
    return input
}

const execAsync = promisify(exec)

const getBiomeBinaryPath = memoize(async () => {
    try {
        const { stdout } = await execAsync('npx --no-install biome --version')
        if (stdout.trim()) {
            return 'npx'
        }
    } catch (error) {
        console.error('Error finding Biome:', error)
    }
    return null
})

export async function formatBiomeFiles(files: string[]) {
    const biomePath = await getBiomeBinaryPath()
    if (!biomePath) {
        console.warn('Biome not found. Skipping formatting.')
        return
    }
    try {
        const command = `${biomePath} biome check --write ${files.join(' ')}`
        const { stdout, stderr } = await execAsync(command)

        if (stderr) {
            console.warn('Biome formatting warning:', stderr)
        }

        console.log('Biome formatting completed:', stdout)
    } catch (error) {
        console.error('Error running Biome format:', error)
    }
}
