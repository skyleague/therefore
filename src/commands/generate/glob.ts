import { Nothing, fstat, isJust, isRight, normalizePath } from '@skyleague/axioms'
import fastGlob from 'fast-glob'

import path from 'node:path'

// should give the same glob behaviour as Prettier
// https://github.com/prettier/prettier/blob/main/src/cli/expand-patterns.js
// also heavily inspired by the logic there :)
export async function expandGlobs({
    patterns,
    ignore = [],
    cwd,
    extension,
}: {
    patterns: string[]
    ignore?: string[]
    cwd: string
    extension: string
}): Promise<string[]> {
    const ignoredDirectories = ['.git', '.svn', '.hg', 'node_modules']
    const globOptions = {
        dot: true,
        ignore: [...ignoredDirectories.map((dir) => `**/${dir}`), ...ignore],
    }

    const entries: string[] = (
        await Promise.allSettled(
            patterns.map(async (pattern) => {
                const absolutePath = path.resolve(cwd, pattern)
                if (ignoredDirectories.some((i) => pattern.includes(i))) {
                    return Nothing
                }

                const eitherStat = await fstat(absolutePath)
                if (isRight(eitherStat) && isJust(eitherStat.right)) {
                    if (eitherStat.right.isFile()) {
                        return fastGlob.escapePath(normalizePath(pattern))
                    }
                    if (eitherStat.right.isDirectory()) {
                        return `${fastGlob.escapePath(normalizePath(path.relative(cwd, absolutePath)))}/**/*${extension}`
                    }

                    return Nothing
                }
                if (pattern.startsWith('!')) {
                    globOptions.ignore.push(normalizePath(pattern.slice(1)))
                    return Nothing
                }
                return normalizePath(pattern)
            }),
        )
    )
        .map((result) => (result.status === 'fulfilled' ? result.value : Nothing))
        .filter(isJust)
    return await fastGlob(entries, globOptions)
}
