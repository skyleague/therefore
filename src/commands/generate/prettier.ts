import type { ThereforeOutputType } from './types.js'

import { Nothing, isNothing } from '@skyleague/axioms'

export async function maybeLoadPrettier() {
    try {
        return await import('prettier')
    } catch {
        return Nothing
    }
}

export async function formatFile(
    prettier: Awaited<ReturnType<typeof maybeLoadPrettier>>,
    input: string,
    file: string,
    type: ThereforeOutputType
): Promise<string> {
    if (isNothing(prettier)) {
        return input
    }
    const options = await prettier.resolveConfig(file)
    return prettier.format(input, {
        ...(options ?? {}),
        ...(type === 'jsonschema' ? { parser: 'json' } : {}),
    })
}
