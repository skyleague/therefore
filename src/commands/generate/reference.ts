import type { FileDefinition, TypescriptDefinition, TypescriptReference } from './types'

import type { ThereforeCst } from '../../lib/primitives/types'
import { toTypescriptDefinition } from '../../lib/visitor/typescript/typescript'

import { evaluate, sha256 } from '@skyleague/axioms'

import path from 'path'

interface ExpandedReference {
    srcPath: string
    targetPath: string
    symbolName: string
    uuid: string
    referenceName: string
    uniqueSymbolName: string
}

const localVersion = new Map<string, number>()
const referenceRegister = new Map<string, string>()

export function requireReference({
    definitions,
    current,
    ref,
    locals,
}: {
    definitions: Record<string, FileDefinition>
    current: FileDefinition
    ref: TypescriptReference
    locals: NonNullable<TypescriptDefinition['locals']>
}): ExpandedReference[] {
    const found = Object.values(definitions)
        .flatMap((d) =>
            d.symbols.map((s) => ({
                uuid: s.definition.uuid,
                symbolName: s.definition.symbolName,
                referenceName: s.definition.referenceName,
                uniqueSymbolName: s.definition.uniqueSymbolName,
                srcPath: d.srcPath,
                targetPath: d.targetPath,
            }))
        )
        .find((s) => s.uuid === ref.uuid)

    if (found === undefined) {
        // this is a locally defined variable
        let sourceSymbol: string
        const reference = evaluate(ref.reference[0])

        const refName = ref.name
        const fileHash = sha256(path.basename(current.srcPath)).slice(0, 4)
        if (refName !== undefined) {
            sourceSymbol = refName
        } else {
            if (!referenceRegister.has(reference.uuid)) {
                if (!localVersion.has(current.targetPath)) {
                    localVersion.set(current.targetPath, 0)
                }
                const localIterator = localVersion.get(current.targetPath)!
                referenceRegister.set(reference.uuid, `local${localIterator}`)
                localVersion.set(current.targetPath, localIterator + 1)
            }
            sourceSymbol = referenceRegister.get(reference.uuid)!
        }

        const { definition } = toTypescriptDefinition({
            sourceSymbol,
            schema: evaluate(reference) as ThereforeCst,
            fileHash,
            exportSymbol: ref.exportSymbol,
            locals,
        })
        current.symbols.push({
            uuid: ref.uuid,
            definition,
            symbolName: sourceSymbol,
            typeOnly: true,
        })

        // make the DFS well defined
        const extra = definition.references.map((r) => requireReference({ definitions, current, ref: r, locals })).flat()
        return [
            ...extra,
            {
                uuid: ref.uuid,
                symbolName: definition.symbolName,
                srcPath: current.srcPath,
                targetPath: current.targetPath,
                referenceName: definition.referenceName,
                uniqueSymbolName: definition.uniqueSymbolName,
            },
        ]
    }
    return [
        {
            uuid: ref.uuid,
            symbolName: found.symbolName,
            srcPath: found.srcPath,
            targetPath: found.targetPath,
            referenceName: found.referenceName,
            uniqueSymbolName: found.uniqueSymbolName,
        },
    ]
}
