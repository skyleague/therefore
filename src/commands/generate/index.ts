import { generate } from './generate'

import { getExtension } from '../../common/template/path'

import type { Argv } from 'yargs'

export function builder(yargs: Argv) {
    return yargs
        .option('files', {
            alias: 'f',
            demandOption: true,
            describe: 'globs to scan for schemas',
            type: 'array',
        })
        .option('ignore-pattern', {
            alias: 'i',
            demandOption: true,
            default: ['**/*.d.ts', 'node_modules'],
            describe: 'globs to exclude',
            type: 'array',
        })
        .option('compile', {
            default: true,
            type: 'boolean',
        })
        .option('ext', {
            default: '.schema.ts',
            type: 'string',
        })
        .option('out-ext', {
            default: '.type.ts',
            type: 'string',
        })
        .option('clean', {
            default: false,
            type: 'boolean',
        })
}

export async function handler(argv: ReturnType<typeof builder>['argv']): Promise<void> {
    const { files = [], 'ignore-pattern': ignorePatterns, ext, 'out-ext': outExt, compile, clean } = await argv

    await generate({
        globs: files.map((f) => f.toString()),
        ignore: ignorePatterns.map((i) => i.toString()),
        extension: ext,
        outputFileRename: (path: string) => path.replace(getExtension(path)!, outExt),
        compile,
        clean,
    })
}

export default {
    command: ['generate', '*'],
    describe: 'generate therefore types',
    builder,
    handler,
}
