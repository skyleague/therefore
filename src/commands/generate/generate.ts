import fs from 'node:fs'
import path from 'node:path'
import { fileURLToPath } from 'node:url'
import { isFailure, isFunction, mapTry } from '@skyleague/axioms'
import type { Schema } from '@typeschema/main'
import { isNode } from '../../lib/cst/cst.js'
import type { Node, SourceNode } from '../../lib/cst/node.js'
import { GenericFileOutput } from '../../lib/output/generic.js'
import type { ThereforeOutput } from '../../lib/output/types.js'
import { TypescriptFileOutput } from '../../lib/output/typescript.js'
import { $ref } from '../../lib/primitives/ref/ref.js'
import { therefore } from '../../lib/primitives/therefore.js'
import { ValidatorType } from '../../lib/primitives/validator/validator.js'
import { autoRef, generateNode } from '../../lib/visitor/prepass/prepass.js'
import { type Prettier, formatBiomeFiles, maybeLoadPrettier } from './format.js'
import { expandGlobs } from './glob.js'

const __dirname = path.dirname(fileURLToPath(import.meta.url))
async function requireModule(module: string): Promise<Record<string, Node | unknown>> {
    const relative = path.relative(__dirname, module).replace(/\\/g, '/')

    const mod = (await import(relative.startsWith('.') ? relative : `./${relative}`)) as Record<string, unknown>
    return (mod.default ?? mod) as Record<string, Node | unknown>
}

export function loadNodes({
    sourceSymbol,
    node,
    sourcePath,
    seen = new Set<string>(),
}: {
    sourceSymbol: string | undefined
    node: Node
    sourcePath: string
    seen?: Set<string>
}): Node[] {
    if (seen.has(node._id)) {
        return []
    }

    const evaluated = therefore.exportSymbol({ symbol: node, sourcePath, sourceSymbol })

    seen.add(evaluated._id)

    // DFS on hypergraph
    const nodes: Node[] = []
    for (const connection of evaluated._connections ?? []) {
        nodes.push(
            ...loadNodes({
                sourceSymbol: undefined,
                node: connection,
                sourcePath: evaluated._sourcePath,
                seen,
            }),
        )
    }
    nodes.push(evaluated)

    return nodes
}

export async function scanModule({
    entry,
    sourcePath,
    basePath: _,
    require = requireModule,
}: {
    entry: string
    sourcePath: string
    basePath: string
    require?: (module: string) => Promise<Record<string, Node | Schema | unknown>>
}) {
    const module = await require(entry)

    const exports: Node[] = []

    for (const [nodeName, nodePromise] of Object.entries(module)) {
        let node = await nodePromise

        if (isFunction(node)) {
            node = await mapTry(node, (x) => x())
        }
        if (!isNode(node)) {
            node = await mapTry(node, (x) => $ref(x as Schema))
            if (isFailure(node)) {
                continue
            }
        }
        const evaluated = loadNodes({ sourceSymbol: nodeName, node: node as Node, sourcePath })

        exports.push(...evaluated)
    }

    console.log(` - found ${[...new Set(exports.map((e) => e._name))].join(', ')}`)

    return exports
}
type SrcPath = string
type ThereforeModules = Record<SrcPath, Node[]>
export async function scanModules({ files, basePath }: { files: string[]; basePath: string }): Promise<ThereforeModules> {
    const modules: ThereforeModules = {}

    const moduleNodes = (
        await Promise.all(
            files.map((entry) => {
                const sourcePath = path.resolve(basePath, entry)
                console.log(`Scanning ${entry}`)

                return scanModule({ entry, basePath, sourcePath })
            }),
        )
    ).flat()

    for (const node of moduleNodes) {
        if (node._sourcePath !== undefined) {
            modules[node._sourcePath] ??= []
            modules[node._sourcePath]?.push(node)
        }
    }

    return modules
}

export function compileModuleExports(modules: ThereforeModules): ThereforeOutput {
    const symbols = Object.values(modules).flatMap((symbols) =>
        symbols.map((s) => {
            if (s instanceof ValidatorType) {
                return s._children[0]
            }
            return s
        }),
    )
    const allSymbols = new WeakSet(symbols)

    for (const symbol of symbols) {
        autoRef(symbol, allSymbols)
    }

    const output: ThereforeOutput = {}
    for (const [_, symbols] of Object.entries(modules)) {
        for (const symbol of symbols) {
            const hasSourcePath = (node: Node): node is SourceNode => '_sourcePath' in node
            if (!hasSourcePath(symbol)) {
                throw new Error('Can only load symbols that have a well defined export path')
            }
            generateNode(symbol)

            for (const generator of therefore.generators) {
                generator.fromSymbol({ symbol, output })
            }
        }
    }
    return output
}

export async function compileOutput(entries: string[], { cwd, prettier = undefined }: { cwd: string; prettier?: Prettier }) {
    therefore.generators = [GenericFileOutput, TypescriptFileOutput]
    therefore.zodCache = new WeakMap()

    const modules = await scanModules({ files: entries, basePath: cwd })

    const output = compileModuleExports(modules)

    const outputFiles: Record<string, string> = {}

    for (const file of Object.values(output)) {
        file.bind()
    }

    for (const [targetPath, file] of Object.entries(output)) {
        const contents = await file.render({ prettier })
        if (contents !== undefined) {
            outputFiles[path.relative(cwd, targetPath)] = contents
        }
    }

    return {
        cleanFn: () => {
            for (const of of Object.values(output)) {
                of.clean()
            }
        },
        outputFiles,
    }
}

export async function generate({
    globs,
    ignore,
    extension,
    clean,
}: {
    globs: string[]
    ignore: string[]
    extension: string
    clean: boolean
}): Promise<void> {
    const cwd = process.cwd()
    const entries = await expandGlobs({ patterns: globs, ignore: [`!${extension}`, ...ignore], cwd, extension })

    const prettier = await maybeLoadPrettier()
    const { cleanFn, outputFiles } = await compileOutput(entries, {
        prettier,
        cwd,
    })

    if (clean) {
        cleanFn()
    }

    for (const [targetPath, contents] of Object.entries(outputFiles)) {
        const absTargetPath = path.join(cwd, targetPath)
        fs.mkdirSync(path.dirname(absTargetPath), { recursive: true })
        fs.writeFileSync(absTargetPath, contents)
    }

    await formatBiomeFiles(Object.keys(outputFiles))
}
