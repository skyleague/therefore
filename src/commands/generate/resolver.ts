import { generatedBy } from './constants.js'
import { requireReference } from './reference.js'
import type { FileDefinition, OutputFile, ReferenceData, TypescriptDefinition } from './types.js'

import packageJson from '../../../package.json' assert { type: 'json' }
import { writeThereforeSchema } from '../../lib/visitor/typescript/typescript.js'
import { createWriter } from '../../lib/writer.js'

import { evaluate, hasPropertiesDefined, unique } from '@skyleague/axioms'

import path from 'node:path'

const { version } = packageJson

export function renderTypescriptSchemaHeader(definition: FileDefinition) {
    const writer = createWriter()
    writer
        .writeLine('/**')
        .writeLine(`* ${generatedBy}@v${version}`)
        .writeLine('* Do not manually touch this')
        .newLineIfLastNot()
        .closeComment()
        .writeLine('/* eslint-disable */')

    for (const importFile of definition.symbols
        .filter(hasPropertiesDefined(['schemaFile']))
        .filter((i) => !i.typeOnly && i.compiledFile === undefined)
        .map((i) => `import ${i.symbolName}Schema from '${i.schemaFile}' assert { type: "json" }`)
        .sort()) {
        writer.writeLine(importFile)
    }

    for (const importFile of [...new Set(definition.symbols.map((i) => i.definition.imports).flat())].sort()) {
        writer.writeLine(importFile)
    }

    for (const importFile of Object.entries(definition.dependencies)
        .map(([targetPath, deps]) => ({
            targetPath,
            deps: `{ ${[...new Set(deps)]
                .sort()
                .map((d) => (definition.dependencyUsesValue[targetPath]?.has(d) ? d : `type ${d}`))
                .join(', ')} }`,
        }))
        .filter(({ targetPath }) => targetPath !== definition.targetPath)
        .map(({ targetPath: f, deps }) => {
            const otherPath = path.relative(path.dirname(definition.targetPath), f).replace('.ts', '').replace(/\\/g, '/')
            return `import ${deps} from '${otherPath.startsWith('.') ? otherPath : `./${otherPath}`}.js'`
        })
        .sort()) {
        writer.writeLine(importFile)
    }

    writer.newLine().newLine()

    return writer.toString()
}

export function renderTypescriptSchemaContent(definition: FileDefinition) {
    const writer = createWriter()
    for (const symbol of Object.values(definition.symbols)) {
        writer.writeLine(symbol.definition.declaration)
        if (symbol.definition.isExported && !symbol.typeOnly) {
            writer.writeLine(
                writeThereforeSchema({
                    uuid: symbol.definition.uuid,
                    schemaReference: `${symbol.symbolName}Schema`,
                    validatorFile: symbol.compiledFile,
                    description: evaluate(symbol.definition.schema).description,
                })
            )
        }
    }

    return writer.toString()
}

export function resolveTypescriptSchema({
    definitions,
    targetPath,
    cwd,
    outputFiles,
    localReferences,
}: {
    definitions: Record<string, FileDefinition>
    targetPath: string
    cwd: string
    outputFiles: OutputFile[]
    localReferences: Record<string, ReferenceData>
}) {
    const definition = definitions[targetPath]!
    const locals: NonNullable<TypescriptDefinition['locals']> = {}
    const required = definition.symbols
        .flatMap((i) =>
            i.definition.references.map((r) => requireReference({ definitions, current: definition, ref: r, locals }))
        )
        .flat(2)
    const uniqueRequires = [...unique(required, (a, b) => a.uuid === b.uuid)]

    const templateContent = renderTypescriptSchemaContent(definition)
    for (const r of uniqueRequires) {
        definition.dependencies[r.targetPath] ??= []
        definition.dependencies[r.targetPath]?.push(r.symbolName)

        if (templateContent.includes(`${r.uuid}:symbolName~value`)) {
            definition.dependencyUsesValue[r.targetPath] ??= new Set()
            definition.dependencyUsesValue[r.targetPath]?.add(r.symbolName)
        }
    }

    for (const symbol of definition.symbols) {
        localReferences[symbol.definition.uuid] ??= {
            symbolName: symbol.definition.symbolName,
            referenceName: symbol.definition.referenceName,
            uniqueSymbolName: symbol.definition.uniqueSymbolName,
            srcPath: targetPath,
        }
    }

    const references: Record<string, ReferenceData> = Object.fromEntries([
        ...required.map(
            (r) =>
                [
                    r.uuid,
                    {
                        symbolName: r.symbolName,
                        uniqueSymbolName: r.uniqueSymbolName,
                        referenceName: r.referenceName,
                        srcPath: r.targetPath,
                    },
                ] as const
        ),
    ])

    const schemaFile = path.relative(cwd, definition.targetPath).replace(/\\/g, '/')
    const relativeSource = `./${path.relative(path.dirname(schemaFile), definition.srcPath).replace(/\.ts$/g, '')}`
    outputFiles.push(
        {
            targetPath,
            relativeSource,
            type: 'typescript',
            template: `${renderTypescriptSchemaHeader(definition)}${templateContent}`,
            data: { ...localReferences, ...references },
            prettify: true,
        },
        ...definition.attachedFiles.map((j) => ({
            targetPath: path.relative(cwd, j.targetPath),
            relativeSource,
            type: j.type,
            template: j.content,
            data: { ...localReferences, ...references },
            prettify: j.prettify,
            clean: j.clean,
        }))
    )
}
