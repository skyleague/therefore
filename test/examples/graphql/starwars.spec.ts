import { compileOutput } from '../../../src/commands/generate/generate.js'
import { expandGlobs } from '../../../src/commands/generate/glob.js'

import { expect, it } from 'vitest'

it('output generation', async () => {
    expect(
        await compileOutput(
            await expandGlobs({ patterns: ['examples/graphql/star-wars/'], extension: '.schema.ts', cwd: process.cwd() }),
            {
                cwd: process.cwd(),
            },
        ),
    ).toMatchSnapshot()
})
