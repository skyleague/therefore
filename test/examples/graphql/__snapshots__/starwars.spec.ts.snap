// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`output generation 1`] = `
{
  "cleanFn": [Function],
  "outputFiles": {
    "examples/graphql/star-wars/query/hero.type.ts": "/**
* Generated by @skyleague/therefore@v1.0.0-local
* Do not manually touch this
*/
/* eslint-disable */

import type { Character } from '../types/character.type.js'
import type { Droid } from '../types/droid.type.js'
import type { Human } from '../types/human.type.js'
import type { CharacterArgs, HeroArgs } from '../types/input.type.js'

export type DroidFieldArgs = CharacterArgs

export type DroidFieldType = Droid

export type DroidField = (args: DroidFieldArgs) => DroidFieldType

export type HeroFieldArgs = HeroArgs

export type HeroFieldType = Character

export type HeroField = (input: HeroFieldArgs) => HeroFieldType

export type HumanFieldArgs = CharacterArgs

export type HumanFieldType = Human

export type HumanField = (args: HumanFieldArgs) => HumanFieldType

",
    "examples/graphql/star-wars/star-wars.graphql": "type Query {
  droid(id: String!): Droid!
  hero(input: HeroArgs!): Character!
  human(id: String!): Human!
}

type Character {
  """Which movies they appear in."""
  appearsIn(id: String!): [Episode!]!

  """The friends of the character, or an empty list if they have none."""
  friends: [Character!]!

  """The id of the character"""
  id: String!

  """The name of the character"""
  name: String!
}

"""A mechanical creature in the Star Wars universe."""
type Droid {
  """Which movies they appear in."""
  appearsIn(id: String!): [Episode!]!

  """The friends of the character, or an empty list if they have none."""
  friends: [Character!]!

  """The id of the character"""
  id: String!

  """The name of the character"""
  name: String!

  """The primary function of the droid."""
  primaryFunction: String!
}

"""One of the films in the Star Wars Trilogy"""
enum Episode {
  EMPIRE
  JEDI
  NEWHOPE
}

input HeroArgs {
  """
  If omitted, returns the hero of the whole saga. If provided, returns the hero of that particular episode.
  """
  episode: Episode!
}

"""A humanoid creature in the Star Wars universe."""
type Human {
  """Which movies they appear in."""
  appearsIn(id: String!): [Episode!]!

  """The friends of the character, or an empty list if they have none."""
  friends: [Character!]!

  """The home planet of the human, or null if unknown."""
  homePlanet: String

  """The id of the character"""
  id: String!

  """The name of the character"""
  name: String!
}",
    "examples/graphql/star-wars/star-wars.type.ts": "/**
* Generated by @skyleague/therefore@v1.0.0-local
* Do not manually touch this
*/
/* eslint-disable */

export const introspection = {"__schema":{"queryType":{"name":"Query"},"mutationType":null,"subscriptionType":null,"types":[{"kind":"OBJECT","name":"Query","fields":[{"name":"droid","type":{"kind":"NON_NULL","ofType":{"kind":"OBJECT","name":"Droid","ofType":null}},"args":[{"name":"id","type":{"kind":"NON_NULL","ofType":{"kind":"SCALAR","name":"Any"}}}]},{"name":"hero","type":{"kind":"NON_NULL","ofType":{"kind":"OBJECT","name":"Character","ofType":null}},"args":[{"name":"input","type":{"kind":"NON_NULL","ofType":{"kind":"SCALAR","name":"Any"}}}]},{"name":"human","type":{"kind":"NON_NULL","ofType":{"kind":"OBJECT","name":"Human","ofType":null}},"args":[{"name":"id","type":{"kind":"NON_NULL","ofType":{"kind":"SCALAR","name":"Any"}}}]}],"interfaces":[]},{"kind":"OBJECT","name":"Character","fields":[{"name":"appearsIn","type":{"kind":"NON_NULL","ofType":{"kind":"LIST","ofType":{"kind":"NON_NULL","ofType":{"kind":"SCALAR","name":"Any"}}}},"args":[{"name":"id","type":{"kind":"NON_NULL","ofType":{"kind":"SCALAR","name":"Any"}}}]},{"name":"friends","type":{"kind":"NON_NULL","ofType":{"kind":"LIST","ofType":{"kind":"NON_NULL","ofType":{"kind":"OBJECT","name":"Character","ofType":null}}}},"args":[]},{"name":"id","type":{"kind":"NON_NULL","ofType":{"kind":"SCALAR","name":"Any"}},"args":[]},{"name":"name","type":{"kind":"NON_NULL","ofType":{"kind":"SCALAR","name":"Any"}},"args":[]}],"interfaces":[]},{"kind":"OBJECT","name":"Droid","fields":[{"name":"appearsIn","type":{"kind":"NON_NULL","ofType":{"kind":"LIST","ofType":{"kind":"NON_NULL","ofType":{"kind":"SCALAR","name":"Any"}}}},"args":[{"name":"id","type":{"kind":"NON_NULL","ofType":{"kind":"SCALAR","name":"Any"}}}]},{"name":"friends","type":{"kind":"NON_NULL","ofType":{"kind":"LIST","ofType":{"kind":"NON_NULL","ofType":{"kind":"OBJECT","name":"Character","ofType":null}}}},"args":[]},{"name":"id","type":{"kind":"NON_NULL","ofType":{"kind":"SCALAR","name":"Any"}},"args":[]},{"name":"name","type":{"kind":"NON_NULL","ofType":{"kind":"SCALAR","name":"Any"}},"args":[]},{"name":"primaryFunction","type":{"kind":"NON_NULL","ofType":{"kind":"SCALAR","name":"Any"}},"args":[]}],"interfaces":[]},{"kind":"OBJECT","name":"Human","fields":[{"name":"appearsIn","type":{"kind":"NON_NULL","ofType":{"kind":"LIST","ofType":{"kind":"NON_NULL","ofType":{"kind":"SCALAR","name":"Any"}}}},"args":[{"name":"id","type":{"kind":"NON_NULL","ofType":{"kind":"SCALAR","name":"Any"}}}]},{"name":"friends","type":{"kind":"NON_NULL","ofType":{"kind":"LIST","ofType":{"kind":"NON_NULL","ofType":{"kind":"OBJECT","name":"Character","ofType":null}}}},"args":[]},{"name":"homePlanet","type":{"kind":"SCALAR","name":"Any"},"args":[]},{"name":"id","type":{"kind":"NON_NULL","ofType":{"kind":"SCALAR","name":"Any"}},"args":[]},{"name":"name","type":{"kind":"NON_NULL","ofType":{"kind":"SCALAR","name":"Any"}},"args":[]}],"interfaces":[]},{"kind":"SCALAR","name":"Any"}],"directives":[]}} as const

",
    "examples/graphql/star-wars/types/character.type.ts": "/**
* Generated by @skyleague/therefore@v1.0.0-local
* Do not manually touch this
*/
/* eslint-disable */

import type { Episode } from './episode.type.js'
import type { CharacterArgs } from './input.type.js'

export type AppearsInResolverArgs = CharacterArgs

export type AppearsInResolverType = (Episode)[]

/**
 * Which movies they appear in.
 */
export type AppearsInResolver = (args: AppearsInResolverArgs) => AppearsInResolverType

export interface Character {
    /**
     * The id of the character
     */
    id: string
    /**
     * The name of the character
     */
    name: string
}

/**
 * The friends of the character, or an empty list if they have none.
 */
export type FriendsResolverType = (Character)[]

/**
 * The friends of the character, or an empty list if they have none.
 */
export type FriendsResolver = () => FriendsResolverType

",
    "examples/graphql/star-wars/types/droid.type.ts": "/**
* Generated by @skyleague/therefore@v1.0.0-local
* Do not manually touch this
*/
/* eslint-disable */

/**
 * A mechanical creature in the Star Wars universe.
 */
export interface Droid {
    /**
     * The id of the character
     */
    id: string
    /**
     * The name of the character
     */
    name: string
}

export type PrimaryFunctionResolverType = string

/**
 * The primary function of the droid.
 */
export type PrimaryFunctionResolver = () => PrimaryFunctionResolverType

",
    "examples/graphql/star-wars/types/episode.type.ts": "/**
* Generated by @skyleague/therefore@v1.0.0-local
* Do not manually touch this
*/
/* eslint-disable */

/**
 * One of the films in the Star Wars Trilogy
 */
export type Episode = 'NEWHOPE' | 'EMPIRE' | 'JEDI'

export type MoreEpisodes = (Episode | 'OTHER')

",
    "examples/graphql/star-wars/types/human.type.ts": "/**
* Generated by @skyleague/therefore@v1.0.0-local
* Do not manually touch this
*/
/* eslint-disable */

export type HomePlanetResolverType = (string | undefined)

/**
 * The home planet of the human, or null if unknown.
 */
export type HomePlanetResolver = () => HomePlanetResolverType

/**
 * A humanoid creature in the Star Wars universe.
 */
export interface Human {
    /**
     * The id of the character
     */
    id: string
    /**
     * The name of the character
     */
    name: string
}

",
    "examples/graphql/star-wars/types/input.type.ts": "/**
* Generated by @skyleague/therefore@v1.0.0-local
* Do not manually touch this
*/
/* eslint-disable */

import type { Episode } from './episode.type.js'

/**
 * The id of the character
 */
export interface CharacterArgs {
    id: string
}

export interface HeroArgs {
    /**
     * If omitted, returns the hero of the whole saga. If provided, returns the hero of that particular episode.
     */
    episode: Episode
}

",
  },
}
`;
